staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/direct/direct.go

```diff
@@ -34,3 +38,29 @@ func Unmarshal(src []byte, dst interface{}) error {
 func Diagnose(src []byte) (string, error) {
 	return modes.Diagnostic.Diagnose(src)
 }
+
+// selfDescribedCBOR is the CBOR encoding of the head of the "self-described CBOR" tag.
+var selfDescribedCBOR = []byte{0xd9, 0xd9, 0xf7}
+
+// Sniff does a constant-time inspection of the provided bytes and returns whether or not they may
+// contain a CBOR data item. If unknown is true, the determination could be wrong.
+func Sniff(src []byte) (cbor bool, unknown bool) {
+	// Objects Outputs from the CBOR encoder will always have this prefix
+	if bytes.HasPrefix(src, selfDescribedCBOR) {
+		return true, true
+	}
+
+	if len(src) > 32 {
+		src = src[:32]
+	}
+	err := modes.DecodeLax.Wellformed(src)
+	switch {
+	case err == nil:
+		return true, false
```

Comment:
I don't know how permissive cbor is (like any random keys you mash are valid perl) ... is it possible this would return a non-error for something that was actually proto or json?

Reply:
It depends. I abandoned the attempt at shared content type detection and split the FieldsV1 content type detection into https://github.com/kubernetes/kubernetes/pull/125790. A valid JSON FieldsV1 can only null or an object. Given that constraint, the first byte is enough to distinguish a valid JSON FieldsV1 from a valid CBOR FieldsV1 without any ambiguity (that's explained in more detail in the linked PR).

A RawExtension isn't necessarily invalid if it contains a JSON number, string, or boolean, which does introduce some potential ambiguous byte sequences that are both valid CBOR and valid JSON. CBOR _does_ specify the "self-described CBOR" tag, which can enclose any CBOR data item without imparting and special semantics and is designed for this situation (https://www.rfc-editor.org/rfc/rfc8949.html#section-3.4.6). It encodes as the prefix 0xd9d9f7, which is not a valid JSON prefix and is also distinct from the Kubernetes Protobuf serializer prefix 0x6b3873.

Now I'm proposing that UnmarshalCBOR ensures Raw is enclosed in the self-described CBOR tag.

https://github.com/kubernetes/kubernetes/pull/125873 shows that field management introduced a de-facto requirement that a RawExtension decoded from a request body may only contain JSON bytes (with the exception https://github.com/kubernetes-sigs/structured-merge-diff/pull/260 seeming to be clearly a bug). I think this constrains the kinds of encodings we can expect to encounter in a persisted RawExtension. In other cases, we already make assumptions about the encoding based on the context.